src/ssl.c

```diff
@@ -5867,6 +5867,136 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
 
     return WOLFSSL_SUCCESS;
 }
+
+void wolfSSL_CertManagerSetVerify(WOLFSSL_CERT_MANAGER* cm, VerifyCallback vc)
+{
+    WOLFSSL_ENTER("wolfSSL_CertManagerSetVerify");
+    if (cm == NULL)
+        return;
+
+    cm->verifyCallback = vc;
+}
+
+static int CertManager_DoVerifyCallback(WOLFSSL_CERT_MANAGER* cm, int ret,
+        DecodedCert* cert)
+{
+
+    int verify_ok = (ret != 0) ? 0 : 1;
+#ifdef WOLFSSL_SMALL_STACK
+    WOLFSSL_X509_STORE_CTX* store;
+    #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    WOLFSSL_X509* x509;
+    #endif
+    char* domain = NULL;
+#else
+    WOLFSSL_X509_STORE_CTX store[1];
+    #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    WOLFSSL_X509           x509[1];
+    #endif
+    char domain[ASN_NAME_MAX];
+#endif
+
+#ifdef WOLFSSL_SMALL_STACK
+    store = (WOLFSSL_X509_STORE_CTX*)XMALLOC(
+        sizeof(WOLFSSL_X509_STORE_CTX), cm->heap, DYNAMIC_TYPE_X509_STORE);
+    if (store == NULL) {
+        return MEMORY_E;
+    }
+    #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    x509 = (WOLFSSL_X509*)XMALLOC(sizeof(WOLFSSL_X509), cm->heap,
+        DYNAMIC_TYPE_X509);
+    if (x509 == NULL) {
+        XFREE(store, cm->heap, DYNAMIC_TYPE_X509);
+        return MEMORY_E;
+    }
+    #endif
+    domain = (char*)XMALLOC(ASN_NAME_MAX, cm->heap, DYNAMIC_TYPE_STRING);
+    if (domain == NULL) {
+        XFREE(store, cm->heap, DYNAMIC_TYPE_X509);
+        #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+        XFREE(x509, cm->heap, DYNAMIC_TYPE_X509);
+        #endif
+        return MEMORY_E;
+    }
+#endif /* WOLFSSL_SMALL_STACK */
+
+    XMEMSET(store, 0, sizeof(WOLFSSL_X509_STORE_CTX));
+#if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    XMEMSET(x509, 0, sizeof(WOLFSSL_X509));
+#endif
+    domain[0] = '\0';
+
+    /* build subject CN as string to return in store */
+    if (cert && cert->subjectCN) {
+        int subjectCNLen = cert->subjectCNLen;
+        if (subjectCNLen > ASN_NAME_MAX-1)
+            subjectCNLen = ASN_NAME_MAX-1;
+        if (subjectCNLen > 0) {
+            XMEMCPY(domain, cert->subjectCN, subjectCNLen);
+            domain[subjectCNLen] = '\0';
+        }
+    }
+
+    store->error = ret;
+    store->error_depth = 1;
+    store->discardSessionCerts = 0;
+    store->domain = domain;
+    store->totalCerts = 1;
```

Comment:
Did you consider having a user supplied `void*` context? Perhaps even just setting `store->userCtx = cm`, so the verify callback can access the certificate manager?

Reply:
I like this approach. The CM pointer could be useful for the callback.